In [1]:
import pandas as pd
import numpy as np
pd.options.mode.chained_assignment = None  

### Question 1: What's the version of NumPy that you installed?

In [3]:
print(np.__version__)

1.21.2


#### answer: 1.21.2

### Question 2: How many records are in the dataset?

In [2]:
car_ds = pd.read_csv('car_price_dataset.csv')

In [7]:
print(car_ds.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

#### Answer: 11914 records

### Question 3: Who are the most frequent car manufacturers (top-3) according to the dataset?

In [9]:
car_ds['Make'].value_counts().nlargest(3)

Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

#### Answer: Chevrolet, Ford, Volkswagen

### Question 4: What's the number of unique Audi car models in the dataset?



In [11]:
print(car_ds['Make'].unique())

['BMW' 'Audi' 'FIAT' 'Mercedes-Benz' 'Chrysler' 'Nissan' 'Volvo' 'Mazda'
 'Mitsubishi' 'Ferrari' 'Alfa Romeo' 'Toyota' 'McLaren' 'Maybach'
 'Pontiac' 'Porsche' 'Saab' 'GMC' 'Hyundai' 'Plymouth' 'Honda'
 'Oldsmobile' 'Suzuki' 'Ford' 'Cadillac' 'Kia' 'Bentley' 'Chevrolet'
 'Dodge' 'Lamborghini' 'Lincoln' 'Subaru' 'Volkswagen' 'Spyker' 'Buick'
 'Acura' 'Rolls-Royce' 'Maserati' 'Lexus' 'Aston Martin' 'Land Rover'
 'Lotus' 'Infiniti' 'Scion' 'Genesis' 'HUMMER' 'Tesla' 'Bugatti']


In [29]:
car_ds[car_ds['Make'] == 'Audi']['Model'].nunique()

34

#### Answer: 34

### Question 5: How many columns in the dataset have missing values?



In [35]:
car_ds.isna().any().sum()

5

#### Answer: 5

### Question 6
1. Find the median value of "Engine Cylinders" column in the dataset.
2. Next, calculate the most frequent value of the same "Engine Cylinders".
3. Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
4. Now, calculate the median value of "Engine Cylinders" once again.

In [37]:
print(car_ds['Engine Cylinders'].mean())

5.628828677213059


In [61]:
cylinder_mode = car_ds['Engine Cylinders'].mode()

In [62]:
print(cylinder_mode)

0    4.0
dtype: float64


In [77]:
print(car_ds.fillna({'Engine Cylinders':cylinder_mode}))

          Make       Model  Year                Engine Fuel Type  Engine HP  \
0          BMW  1 Series M  2011     premium unleaded (required)      335.0   
1          BMW    1 Series  2011     premium unleaded (required)      300.0   
2          BMW    1 Series  2011     premium unleaded (required)      300.0   
3          BMW    1 Series  2011     premium unleaded (required)      230.0   
4          BMW    1 Series  2011     premium unleaded (required)      230.0   
...        ...         ...   ...                             ...        ...   
11909    Acura         ZDX  2012     premium unleaded (required)      300.0   
11910    Acura         ZDX  2012     premium unleaded (required)      300.0   
11911    Acura         ZDX  2012     premium unleaded (required)      300.0   
11912    Acura         ZDX  2013  premium unleaded (recommended)      300.0   
11913  Lincoln      Zephyr  2006                regular unleaded      221.0   

       Engine Cylinders Transmission Type      Driv

In [90]:
car_ds['Engine Cylinders'].fillna(cylinder_mode, inplace=True)

In [91]:
car_ds[car_ds['Engine Cylinders'].isna()==True]

Make    Model  Year             Engine Fuel Type  Engine HP  \
1983   Chevrolet  Bolt EV  2017                     electric      200.0   
1984   Chevrolet  Bolt EV  2017                     electric      200.0   
3716  Volkswagen   e-Golf  2015                     electric      115.0   
3717  Volkswagen   e-Golf  2015                     electric      115.0   
3718  Volkswagen   e-Golf  2016                     electric      115.0   
3719  Volkswagen   e-Golf  2016                     electric      115.0   
5778  Mitsubishi   i-MiEV  2014                     electric        NaN   
5779  Mitsubishi   i-MiEV  2016                     electric       66.0   
5780  Mitsubishi   i-MiEV  2017                     electric       66.0   
8373      Toyota  RAV4 EV  2012                     electric      154.0   
8695       Mazda     RX-7  1993             regular unleaded      255.0   
8696       Mazda     RX-7  1994             regular unleaded      255.0   
8697       Mazda     RX-7  1995             regular unleaded      255.0   
8698       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8699       Mazda     RX-8  2009  premium unleaded (required)      212.0   
8700       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8701       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8702       Mazda     RX-8  2009  premium unleaded (required)      232.0   
8703       Mazda     RX-8  2009  premium unleaded (required)      212.0   
8704       Mazda     RX-8  2009  premium unleaded (required)      212.0   
8705       Mazda     RX-8  2010  premium unleaded (required)      232.0   
8706       Mazda     RX-8  2010  premium unleaded (required)      212.0   
8707       Mazda     RX-8  2010  premium unleaded (required)      212.0   
8708       Mazda     RX-8  2010  premium unleaded (required)      232.0   
8709       Mazda     RX-8  2010  premium unleaded (required)      232.0   
8710       Mazda     RX-8  2011  premium unleaded (required)      212.0   
8711       Mazda     RX-8  2011  premium unleaded (required)      232.0   
8712       Mazda     RX-8  2011  premium unleaded (required)      232.0   
8713       Mazda     RX-8  2011  premium unleaded (required)      232.0   
8714       Mazda     RX-8  2011  premium unleaded (required)      212.0   

      Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
1983               NaN      DIRECT_DRIVE  front wheel drive              4.0   
1984               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3716               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3717               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3718               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3719               NaN      DIRECT_DRIVE  front wheel drive              4.0   
5778               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
5779               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
5780               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
8373               NaN      DIRECT_DRIVE  front wheel drive              4.0   
8695               NaN            MANUAL   rear wheel drive              2.0   
8696               NaN            MANUAL   rear wheel drive              2.0   
8697               NaN            MANUAL   rear wheel drive              2.0   
8698               NaN            MANUAL   rear wheel drive              4.0   
8699               NaN         AUTOMATIC   rear wheel drive              4.0   
8700               NaN            MANUAL   rear wheel drive              4.0   
8701               NaN            MANUAL   rear wheel drive              4.0   
8702               NaN            MANUAL   rear wheel drive              4.0   
8703               NaN         AUTOMATIC   rear wheel drive              4.0   
8704               NaN         AUTOMATIC   rear wheel drive              4.0   
8705  

#### Answer:
1. 5.628828677213059
2. 4.0
3. car_ds['Engine Cylinders'].fillna(cylinder_mode, inplace=True)
4. No

### Question 7: How many columns in the dataset have missing values?
1. Select all the "Lotus" cars from the dataset.
2. Select only columns "Engine HP", "Engine Cylinders".
3. Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9 What's the value of the first element of w



In [3]:
lotus_df = car_ds[car_ds.Make=='Lotus']

In [103]:
print(lotus_df.head())

       Make  Model  Year             Engine Fuel Type  Engine HP  \
3912  Lotus  Elise  2009  premium unleaded (required)      189.0   
3913  Lotus  Elise  2009  premium unleaded (required)      218.0   
3914  Lotus  Elise  2009  premium unleaded (required)      189.0   
3915  Lotus  Elise  2010  premium unleaded (required)      189.0   
3916  Lotus  Elise  2010  premium unleaded (required)      218.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
3912  Exotic,High-Performance      Compact 

In [4]:
hp_cylinders = lotus_df[['Engine HP','Engine Cylinders']]

In [5]:
hp_cylinders.drop_duplicates(inplace=True)

In [6]:
X = hp_cylinders.to_numpy()

In [7]:
print(X)

[[189.   4.]
 [218.   4.]
 [217.   4.]
 [350.   8.]
 [400.   6.]
 [276.   6.]
 [345.   6.]
 [257.   4.]
 [240.   4.]]


In [8]:
XT = np.transpose(X)

In [9]:
XTX = np.dot(XT,X)

In [10]:
print(XTX)

[[7.31684e+05 1.34100e+04]
 [1.34100e+04 2.52000e+02]]


In [11]:
inversed_X = np.linalg.inv(XTX)

In [12]:
print(inversed_X)

[[ 5.53084235e-05 -2.94319825e-03]
 [-2.94319825e-03  1.60588447e-01]]


In [14]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

In [15]:
w = np.dot(np.dot(inversed_X,XT),y)

In [16]:
print(w)

[  4.59494481 -63.56432501]


#### Anwer:
8. What's the value of the first element of w: 4.59494481